internal/gamepad/gamepad.go

```diff
@@ -187,11 +187,30 @@ type Gamepad struct {
 	native nativeGamepad
 }
 
+type mappingKind int
+
+const (
+	mappingKindNone mappingKind = iota
+	mappingKindAxis
+	mappingKindButtonAnalog
+	mappingKindButtonDigital
+	mappingKindButtonMixed
+	mappingKindHatUp
+	mappingKindHatRight
+	mappingKindHatDown
+	mappingKindHatLeft
```

Comment:
For a standard map, there are only axes and buttons. Do we need some types of buttons? Also, do we need hats here?

Reply:
See https://github.com/hajimehoshi/ebiten/pull/2587 - the mappingKind is what the _source_ is.

E.g. shoulder buttons can be represented by any of an axis, hat or button.

Reply:
Similarly, the split of analog/digital source buttons is needed to support as Linux only supports digital buttons but some other platforms (and Ebitengine's own API) support analog ones.

Reply:
(Could ditch THAT split though by deprecating internal isButtonPressed APIs and only using buttonValue - which I didn't do, as on some platforms they both go through different OS APIs, letting the OS implement its own threshold)

Reply:
So this is for Linux and is an implementation details.

I think we should return an interface (or a struct with methods) to hide the logic, rather than exposing some various key types.

Reply:
Not just Linux - every implementation has its own issues there. I do not think it makes sense to have each OS's implementation define its own code for this, as all standard layout items should really be backed by a native button, axis or hat.

I could however change the enum into a globally defined interface + globally defined structs that are based on a native button, axis or hat.

Reply:
Also, it _seems_ like mappingKindButtonAnalog is never needed - as long as all OSes that implement buttonValue also implement isButtonPressed, I can ditch that one.

Reply:
Similarly, if you are OK with having all OSes that implement isButtonPressed also implement buttonValue (even if that only ever returns 0.0 or 1.0), we can move to only one button type.

Reply:
To elaborate - I think it'd be a good policy decision that any standard layout item MUST be backed by a native button, axis or hat, so that the StandardGamepad* APIs can't return anything that GamepadAxisValue/IsGamepadButtonPressed cannot return as well. That is why I really would like to require that this is not an OS-specific interface but there's really a requirement that the input shall be reachable by existing APIs.

Reply:
> Not just Linux - every implementation has its own issues there. I do not think it makes sense to have each OS's implementation define its own code for this, as all standard layout items should really be backed by a native button, axis or hat.
> I could however change the enum into a globally defined interface + globally defined structs that are based on a native button, axis or hat.

Hmm I see. As long as the interface has the notion of hats, we can go with your PR. We can revisit this (how to refactor this) later.

> Also, it seems like mappingKindButtonAnalog is never needed - as long as all OSes that implement buttonValue also implement isButtonPressed, I can ditch that one.
> Similarly, if you are OK with having all OSes that implement isButtonPressed also implement buttonValue (even if that only ever returns 0.0 or 1.0), we can move to only one button type.

I think I'm fine to do this. Let's remove isButtonPressed, which might be redundant.

Reply:
Isn't it inevitable to have types for all four directions for hats?

Reply:
> Hmm I see. As long as the nativeGamepad interface has the notion of hats, we can go with your PR. We can revisit this (how to refactor this) later.

If you have already done with interfaces, that's also fine.

Reply:
Did the global interface way now (which also unified hats into one type). I am not sure about isButtonPressed, though.

gamepad_js.go, gamepad_nintendosdk.go and  implements buttonValue and isButtonPressed separately. As I do not have access to the Nintendo SDK, I can't judge whether that is important.

Reply:
> Did the global interface way now (which also unified hats into one type). I am not sure about isButtonPressed, though.
> gamepad_js.go, gamepad_nintendosdk.go implement buttonValue and isButtonPressed separately. As I do not have access to the Nintendo SDK, I can't judge whether that is important.

Ah I see. As at least *_js has a separate API for is-pressed and value, let's keep the is-pressed part. Sorry for going back and forth.

Reply:
For gamepad_js, I'd slightly tend towards keeping the separation, as that means the thresholding is implemented in the web browser. Although truth is, browsers are most likely rather dumb about this...

So I guess the decision boils down to gamepad_nintendosdk. If anyone may have better thresholding than just a simple > x comparison (e.g. dynamic thresholding based on the gamepad's specs or past performance), then that'd be Nintendo.

Reply:
Actually, let's do this: I'll ONLY keep mixed buttons, and rather replace all the panics in .buttonValue by if-pressed-then-1-else-0 logic.

Reply:
Done that now. All the buttonValue panics are gone, so we no longer need more than one button input type.

Reply:
FYI: For NintendoSwitch, there are no analog button and a button value only takes 0 and 1